File: DynamicKey/AgoraDynamicKey/ruby/lib/dynamic_key2/stt_token_builder.rb
module AgoraDynamicKey2
  class SttTokenBuilder
    # Build a Token007 that carries RTC, RTM, and STT services.
    #
    # app_id: The App ID issued to you by Agora.
    # app_certificate: Certificate of the application that you registered in the Agora Dashboard.
    # channel_name: Unique channel name for the AgoraRTC session in the string format.
    # rtc_account: The RTC user's account, max length is 255 Bytes.
    # rtc_role: ROLE_PUBLISHER: A broadcaster/host in a live-broadcast profile.
    #     ROLE_SUBSCRIBER: An audience member in a live-broadcast profile.
    # rtc_token_expire: represented by the number of seconds elapsed since now.
    # join_channel_privilege_expire: represented by the number of seconds elapsed since now.
    # pub_audio_privilege_expire: represented by the number of seconds elapsed since now.
    # pub_video_privilege_expire: represented by the number of seconds elapsed since now.
    # pub_data_stream_privilege_expire: represented by the number of seconds elapsed since now.
    # rtm_user_id: The RTM user's account, max length is 255 Bytes.
    # rtm_token_expire: represented by the number of seconds elapsed since now.
    # return: The RTC, RTM, and STT token.
    def self.build_token(app_id, app_certificate, channel_name, rtc_account, rtc_role, rtc_token_expire,
                         join_channel_privilege_expire, pub_audio_privilege_expire, pub_video_privilege_expire,
                         pub_data_stream_privilege_expire, rtm_user_id, rtm_token_expire)
      access_token = AgoraDynamicKey2::AccessToken.new(app_id, app_certificate, rtc_token_expire)
      service_rtc = AgoraDynamicKey2::ServiceRtc.new(channel_name, rtc_account)

      service_rtc.add_privilege(AgoraDynamicKey2::ServiceRtc::PRIVILEGE_JOIN_CHANNEL, join_channel_privilege_expire)
      if rtc_role == AgoraDynamicKey2::RtcTokenBuilder::ROLE_PUBLISHER
        service_rtc.add_privilege(AgoraDynamicKey2::ServiceRtc::PRIVILEGE_PUBLISH_AUDIO_STREAM, pub_audio_privilege_expire)
        service_rtc.add_privilege(AgoraDynamicKey2::ServiceRtc::PRIVILEGE_PUBLISH_VIDEO_STREAM, pub_video_privilege_expire)
        service_rtc.add_privilege(AgoraDynamicKey2::ServiceRtc::PRIVILEGE_PUBLISH_DATA_STREAM, pub_data_stream_privilege_expire)
      end
      access_token.add_service(service_rtc)

      service_rtm = AgoraDynamicKey2::ServiceRtm.new(rtm_user_id)
      service_rtm.add_privilege(AgoraDynamicKey2::ServiceRtm::PRIVILEGE_JOIN_LOGIN, rtm_token_expire)
      access_token.add_service(service_rtm)

      access_token.add_service(AgoraDynamicKey2::ServiceStt.new)
      access_token.build
    end
  end
end

File: DynamicKey/AgoraDynamicKey/go/src/stttokenbuilder/stttokenbuilder.go
package stttokenbuilder

import (
	accesstoken "github.com/AgoraIO/Tools/DynamicKey/AgoraDynamicKey/go/src/accesstoken2"
	rtctokenbuilder "github.com/AgoraIO/Tools/DynamicKey/AgoraDynamicKey/go/src/rtctokenbuilder2"
)

// BuildToken builds a Token007 that carries RTC, RTM, and STT services.
//
// @param appId: The App ID issued to you by Agora. Apply for a new App ID from
// Agora Dashboard if it is missing from your kit. See Get an App ID.
// @param appCertificate: Certificate of the application that you registered in
// the Agora Dashboard. See Get an App Certificate.
// @param channelName: Unique channel name for the AgoraRTC session in the string format.
// @param rtcAccount: The RTC user's account, max length is 255 Bytes.
// @param rtcRole: RolePublisher: A broadcaster or host in a live-broadcast profile.
// RoleSubscriber: An audience member in a live-broadcast profile.
// @param rtcTokenExpire: represented by the number of seconds elapsed since now. If, for example,
// you want to access the Agora Service within 10 minutes after the token is generated,
// set rtcTokenExpire as 600(seconds).
// @param joinChannelPrivilegeExpire: represented by the number of seconds elapsed since now.
// If, for example, you want to join channel and expect stay in the channel for 10 minutes,
// set joinChannelPrivilegeExpire as 600(seconds).
// @param pubAudioPrivilegeExpire: represented by the number of seconds elapsed since now.
// If, for example, you want to enable publish audio privilege for 10 minutes,
// set pubAudioPrivilegeExpire as 600(seconds).
// @param pubVideoPrivilegeExpire: represented by the number of seconds elapsed since now.
// If, for example, you want to enable publish video privilege for 10 minutes,
// set pubVideoPrivilegeExpire as 600(seconds).
// @param pubDataStreamPrivilegeExpire: represented by the number of seconds elapsed since now.
// If, for example, you want to enable publish data stream privilege for 10 minutes,
// set pubDataStreamPrivilegeExpire as 600(seconds).
// @param rtmUserId: The RTM user's account, max length is 255 Bytes.
// @param rtmTokenExpire: represented by the number of seconds elapsed since now. If, for example,
// you want to access the Agora Service within 10 minutes after the token is generated,
// set rtmTokenExpire as 600(seconds).
//
// return The RTC, RTM, and STT token.
func BuildToken(appId string, appCertificate string, channelName string, rtcAccount string, rtcRole rtctokenbuilder.Role, rtcTokenExpire uint32,
	joinChannelPrivilegeExpire uint32, pubAudioPrivilegeExpire uint32, pubVideoPrivilegeExpire uint32, pubDataStreamPrivilegeExpire uint32,
	rtmUserId string, rtmTokenExpire uint32) (string, error) {
	token := accesstoken.NewAccessToken(appId, appCertificate, rtcTokenExpire)

	serviceRtc := accesstoken.NewServiceRtc(channelName, rtcAccount)
	serviceRtc.AddPrivilege(accesstoken.PrivilegeJoinChannel, joinChannelPrivilegeExpire)
	if rtcRole == rtctokenbuilder.RolePublisher {
		serviceRtc.AddPrivilege(accesstoken.PrivilegePublishAudioStream, pubAudioPrivilegeExpire)
		serviceRtc.AddPrivilege(accesstoken.PrivilegePublishVideoStream, pubVideoPrivilegeExpire)
		serviceRtc.AddPrivilege(accesstoken.PrivilegePublishDataStream, pubDataStreamPrivilegeExpire)
	}
	token.AddService(serviceRtc)

	serviceRtm := accesstoken.NewServiceRtm(rtmUserId)
	serviceRtm.AddPrivilege(accesstoken.PrivilegeLogin, rtmTokenExpire)
	token.AddService(serviceRtm)

	token.AddService(accesstoken.NewServiceStt())

	return token.Build()
}
